cluster_management/src/main/java/com/pinterest/rocksplicator/BulkIngestionStateModelFactory.java

```diff
@@ -0,0 +1,150 @@
+      this.resourceName = resourceName;
+      this.partitionName = partitionName;
+      this.host = host;
+      this.adminPort = adminPort;
+      this.cluster = cluster;
+    }
+
+    /**
+     * Callback for OFFLINE to ONLINE transition.
+     * This can happen when 1) a new partition is first launched on a host; or 2) an existing
+     * partition is reopened when restart/zk temporarily expires.
+     * We first make sure the DB is open and then load SST files into the DB.
+     */
+    @Override
+    public void onBecomeOnlineFromOffline(Message message, NotificationContext context) {
+      super.checkSanity("OFFLINE", "ONLINE", message);
+      Utils.logTransitionMessage(message);
+      Utils.addDB(Utils.getDbName(partitionName), adminPort);
+
+      try {
+        String dbName = Utils.getDbName(partitionName);
+
+        // open the DB if it's currently not opened yet
+        Utils.addDB(dbName, adminPort);
+
+        // see if the DB needs to be rebuilt. If there is anything in the meta,
+        // assume that we have a complete local replica. We can ensure this
+        // by locking the helix shard config during data uploads
+        CheckDBResponse localStatus = Utils.checkLocalDB(dbName, adminPort);
+
+        // check that the metadata is set and well formed, otherwise we
+        // build a replica from another host serving.
+        boolean needRebuild = !localStatus.isSetDb_meta_data();
+        needRebuild = needRebuild && !localStatus.getDb_meta_data().isSetS3_bucket();
+        needRebuild = needRebuild && !localStatus.getDb_meta_data().isSetS3_path();
+        if (needRebuild) {
+          HelixAdmin admin = context.getManager().getClusterManagmentTool();
+          ExternalView view = admin.getResourceExternalView(cluster, resourceName);
+          Map<String, String> stateMap = view.getStateMap(partitionName);
+          // find live replicas
+
+          for (Map.Entry<String, String> instanceNameAndRole : stateMap.entrySet()) {
+            String role = instanceNameAndRole.getValue();
+            if (!role.equalsIgnoreCase("ONLINE")) {
+              continue;
+            }
+
+            String hostPort = instanceNameAndRole.getKey();
+            String host = hostPort.split("_")[0];
+            int port = Integer.parseInt(hostPort.split("_")[1]);
+
+            if (this.host.equals(host)) {
+              // myself
+              continue;
+            }
+
+            // We've got an online host that has the partition. Copy the data
+            String hdfsPath = "/rocksplicator/" + cluster + "/" + dbName + "/" + hostPort + "/"
+                + String.valueOf(System.currentTimeMillis());
+
+            // backup a snapshot from the upstream host, and restore it locally
+            LOG.error("Backup " + dbName + " from " + hostPort);
+            Utils.backupDB(host, port, dbName, hdfsPath);
+            LOG.error("Restore " + dbName + " from " + hdfsPath);
+            Utils.closeDB(dbName, adminPort);
```

Comment:
if either backup or restore throws an exception do we want to try using other hosts to copy the data ? in the current code we will go to the catch block and exit.